Libraries/LibGC/Heap+Swift.swift

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2024, Andrew Kaster <andrew@ladybird.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+import AK
+@_exported import GCCxx
+
+extension GC.Heap {
+    public func withDeferredGC<R, E>(_ body: () throws(E) -> R) throws(E) -> R {
+        self.defer_gc()
+        defer { self.undefer_gc() }
+        return try body()
+    }
+}
+
+// FIXME: Cell and Cell::Visitor are not imported properly, so we have to treat them as OpaquePointer
+public protocol HeapAllocatable {
+    static func allocate(on heap: GC.Heap) -> UnsafeMutablePointer<Self>
+
+    init(cell: OpaquePointer)
+
+    func finalize()
+    func visitEdges(_ visitor: OpaquePointer)
+
+    var cell: OpaquePointer { get }
+}
+
+// FIXME: Figure out why other modules can't conform to HeapAllocatable
+public struct HeapString: HeapAllocatable {
+    public var string: Swift.String
+
+    public init(cell: OpaquePointer) {
+        string = "Hello, World!"
+        self.cell = cell
+    }
+
+    // FIXME: HeapAllocatable cannot be exposed to C++ yet, so we're off to void* paradise
+    public static func create(on heap: GC.Heap) -> OpaquePointer {
+        precondition(heap.is_gc_deferred())
+        return allocate(on: heap).pointee.cell
+    }
+
+    public var cell: OpaquePointer
+}
+
+// Here be dragons
+
+extension GC.ForeignCell.ForeignClass {
+    fileprivate var type: any HeapAllocatable.Type {
+        let typeObject = Unmanaged<AnyObject>.fromOpaque(ptr).takeUnretainedValue()
+        guard let type = typeObject as? any HeapAllocatable.Type else {
+            fatalError("Passed foreign class \(self.class_name) but it wasn't a Swift type!")
+        }
+        return type
+    }
+
+    fileprivate init(from type: any HeapAllocatable.Type) {
+        self.init()
+        self.class_name = AK.String(swiftString: Swift.String(describing: type))
+        self.ptr = Unmanaged.passRetained(type as AnyObject).toOpaque()
```

Comment:
The "type as AnyObject" here is relying on an aspect of Objective-C interoperability that won't be portable. I'd suggest that what you want is to coerce the type to `Any.Type`, which represents any type in Swift. It is guaranteed to be a single pointer (the "type metadata pointer", if you're familiar with the Swift runtime terminology) that is immortal. Given that this is very low-level code that is going to get a lot of scrutiny and not a lot of changes, I think it's fine to use `unsafeBitCast(_:to:)` to map between the `Any.Type` and the pointer you're getting on the C++ side, e.g.,

```suggestion
        self.ptr = asTypeMetadataPointer(type)
```

where `asTypeMetadataPointer` is:

```swift
func asTypeMetadataPointer(_ type: Any.Type) -> UnsafeRawMutablePointer {
  unsafeBitCast(type, to: UnsafeRawMutablePointer.self)
}
```

Reply:
Funnily enough I made the same mistake and had to learn about it the hard way in swift-java 😄